### DISCUSSION TITLE
How to use DEBUG=1 and COMPARE=<filename> to debug the matching results of your engine

### DISCUSSION THREAD

**Discussion by mjethani** [ghostery/adblocker#2165] [Show and tell]
Suppose you are working on the Cliqz ad blocking engine and you want to see if your changes are causing any false positives or any false negatives.

You can start by getting the debug output of another engine that you think gives correct results. This could be a previous stable version of Cliqz, or it could be a different engine like uBlock Origin. Let's take uBlock Origin as an example:

```console
$ make ublock DEBUG=1
./ublock.debug.json
```

The above command should produce a file called `ublock.debug.json`.

Now you can compare the results from Cliqz with those from uBlock Origin:

```console
$ make cliqz DEBUG=1 COMPARE=ublock.debug.json
./cliqz.diff.json
```

This should produce a much smaller file called `cliqz.diff.json`.

An entry in the file may look like this:

```json
{
  "index": 49931,
  "url": "https://src.ebay-us.com/800u3jujhrtna4oy.js?t2h0syoxz29o70at=usllpic0&m4g3fbcsoflgkza2=31ceded617b0a69c2824c2d9fffd7bf3",
  "frameUrl": "https://pages.ebay.com/identity/device/t_n6.html?org_id=usllpic0&session_id=31ceded617b0a69c2824c2d9fffd7bf3",
  "cpt": "script",
  "match": false,
  "matchDebug": null,
  "compareMatchDebug": "||src.ebay-us.com/*=usllpic$script,domain=ebay.com",
  "kind": "false -ve"
}
```

What the value of the `kind` field tells us is that we have a [false negative](https://en.wikipedia.org/wiki/False_positives_and_false_negatives). `compareMatchDebug` tells us the filter that was matched by uBlock Origin but not by Cliqz. `matchDebug` tells us what filter Cliqz matched, if any. Even though it's clearly a false negative, in some cases it can be so due to an incorrectly matched [exception](https://help.eyeo.com/adblockplus/how-to-write-filters#allowlist).

And that's all there is to it.

The new `DEBUG=1` and `COMPARE=<filename>` options should make development faster by making mistakes easier to catch and analyze.

**Comment by mjethani** [comment]
@remusao @ameshkov @antonok-edm @gorhill 

Replace Cliqz above with tsurlfilter, Brave, or uBlock Origin and you've got a way to find and analyze issues more quickly.

I would compare results against either Adblock Plus or uBlock Origin since they are both extremely stable and most correct in their implementations.

**Comment by ameshkov** [comment]
Okay, I don't get what's wrong, check this out:

`make tsurlfilter DEBUG=1 COMPARE=adblockplus.debug.json`

All the differences look like this:
```
  {
    "index": 9796,
    "url": "https://www.googletagmanager.com/gtm.js?id=GTM-WJQWWTD",
    "frameUrl": "http://www.spiegel.de/video/theresa-may-gibt-brexit-erklaerung-ab-livestream-video-99024890.html",
    "cpt": "script",
    "match": false,
    "matchDebug": "@@||spiegel.de^$genericblock,generichide",
    "compareMatchDebug": "/gtm.js",
    "kind": "false -ve"
  },
```

The question is why neither adblockplus nor ublock apply `genericblock` here?
In theory, this `$genericblock` should've unblocked GTM in their case, but it didn't happen.

**Reply by ameshkov** [reply]
Okay, `make clean ; npm install` changes things a little since the requests list is different, but the issue with `$genericblock` stays the same.

`make tsurlfilter DEBUG=1 COMPARE=ublock.debug.json`

Here are some requests that should've been unblocked by the `$genericblock` rule but weren't:
```
  {
    "index": 19966,
    "url": "https://sams.spiegel.de/id?d_visid_ver=5.2.0&d_fieldgroup=A&mcorgid=79655FCF5C1D42160A495E15%40AdobeOrg&mid=16836366873838520273040683356840932679&d_coppa=true&ts=1628627559223",
    "frameUrl": "https://www.spiegel.de/consent-a-?targetUrl=https%3A%2F%2Fwww.spiegel.de%2F",
    "cpt": "xhr",
    "match": false,
    "matchDebug": "@@||spiegel.de^$genericblock,generichide",
    "compareMatchDebug": "||sams.spiegel.de^",
    "kind": "false -ve"
  },
  {
    "index": 19967,
    "url": "https://assets.adobedtm.com/extensions/EPbde2f7ca14e540399dcc1f8208860b7b/AppMeasurement.min.js",
    "frameUrl": "https://www.spiegel.de/consent-a-?targetUrl=https%3A%2F%2Fwww.spiegel.de%2F",
    "cpt": "script",
    "match": false,
    "matchDebug": "@@||spiegel.de^$genericblock,generichide",
    "compareMatchDebug": "||adobedtm.com^$3p,domain=~adobe.com",
    "kind": "false -ve"
  },
```

This one was not blocked by ublock although it looks like it should've been:
```
  {
    "index": 10021,
    "url": "https://www.google-analytics.com/j/collect?v=1&_v=j92&a=393452224&t=pageview&_s=1&dl=https%3A%2F%2Fthechive.com%2F&ul=en-us&de=UTF-8&dt=An%20improbable%20hall%20of%20fame%20journey%20and%20the%201.000%20batting%20average%20(8%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Actors%20you%20WON%E2%80%99T%20BELIEVE%20are%20turning%2060%20this%20year%20(16%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Animals%20prove%20you%20can%20find%20friends%20in%20the%20most%20unlikely%20of%20places%20(22%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09A%20Bikini%20a%20day%20keeps%20the%20Bikini%20Doctor%20away%E2%80%A6%20right%3F%20(35%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Daily%20Morning%20Awesomeness%20(40%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Turns%20out%20marriage%20is%20a%20big%20f%40%23%25ing%20joke%20(31%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Kids%20bring%20some%20crazy%20sh*t%20to%20show%20and%20tell%20(19%20GIFs)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Blazing%20hot%20Dutch%20runner%20Lieke%20Klaver%20won%20the%20Olympics%20(21%20photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Quentin%20Tarantino%20explains%20why%20he%E2%80%99ll%20NEVER%20give%20his%20mom%20a%20cent%20of%20his%20film%20money%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Daily%20Afternoon%20Randomness%20(49%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Keep%20calm%20and%20tug%20on%20(38%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Some%20small%20hills%20are%20worth%20dying%20on%2C%20just%20ask%20these%20folks%20(20%20GIFs)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09The%20best%20tweets%20from%20parents%20of%20all%20time%20last%20week%20(25%20photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Folks%20share%20memories%20of%20their%20fathers%20who%20passed%20away%20(21%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Daily%20Morning%20Awesomeness%20(38%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09What%E2%80%99s%20the%20Worst%20Pain%20You%E2%80%99ve%20Ever%20Felt%3F%20(15%20GIFs)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09All%20systems%20GO!%20Prepare%20for%20Liftoff%20in%203%E2%80%A62%E2%80%A61%E2%80%A6%20(45%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Read%20Between%20the%20Tan%20Lines%20(35%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09My%20sweet%20Big%20Mouth%20Billy%20Bass%2C%20look%20what%20they%20did%20to%20you!%20(Video)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Weekend%20Afternoon%20Randomness%20(36%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09GIFs%20prove%3A%20Can%E2%80%99t%20out-Random%20the%20internet%20(25%20GIFs)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09These%20male%20authors%20prove%20they%20know%20nothing%20about%20women%E2%80%99s%20bodies%20(31%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Nice%20Design%20Work%20Ya%20Absolute%20Prick!%20(23%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Upon%20closer%20inspection%20HOLY%20S***!%20(35%20Photos)%09%09%09%0A%09%09%0A%09%09%09%0A%09%09%09%09Weekend%20Morning%20Awesomeness%20(37%20photos)&sd=24-bit&sr=800x600&vp=800x600&je=0&_u=YEBAAAABAAAAAC~&jid=331000972&gjid=1930197999&cid=827252379.1628628622&tid=UA-6481362-1&_gid=1175243576.1628628622&_r=1&gtm=2wg891P83WPK&cg2=homepage&cd2=&cd3=homepage&cd4=&cd5=&cd13=C_JS%3A%5Bobject%20Object%5Dundefined%2CmetaId%3Ano%2CbodyPostId%3Ano&z=388502388",
    "frameUrl": "https://thechive.com/",
    "cpt": "xhr",
    "match": true,
    "matchDebug": "/collect?*&cid=",
    "compareMatchDebug": null,
    "kind": "false +ve"
  },
  ```

@gorhill could you please check this out?

Also, there's a bunch of legit tsurlfilter mistakes in the diff so it's super helpful, thank you @mjethani!

**Reply by mjethani** [reply]
> The question is why neither adblockplus nor ublock apply `genericblock` here?

This could become a discussion of its own: https://github.com/cliqz-oss/adblocker/discussions/2170

**Comment by gorhill** [comment]
> requests that should've been unblocked by the $genericblock rule but weren't.

uBO does not support `genericblock`, by design, see <https://github.com/gorhill/uBlock/issues/967#issuecomment-159363578>.

That option solely exists to work around anti-blockers, and it's essentially a directive to to tell the blocker to surrender by blocking less. uBO's approach has been to defuse the anti-blockers (using scriptlets, redirect rules), rather than not blocking to a point which satisfies the anti-blockers.

---

> This one was not blocked by ublock although it looks like it should've been:

Probably same as <https://github.com/cliqz-oss/adblocker/pull/2125#issuecomment-896699694>.

**Reply by ameshkov** [reply]
Got it, explains the difference.

On a side note, the decision sounds strange to me. List maintainers still can unblock whatever they need using the regular `@@||example.org^$domain=domain.com` syntax, `$genericblock` is just a convenient shortcut for that.

**Reply by gorhill** [reply]
If we give them the tools to work around anti-blockers, they will use them rather than fall back to using exception filters. My point is not that I do not trust filter list maintainers, my point was to give them the tools to work around anti-blockers without having to "surrender" to the dictates of sites -- I do understand that sometimes there is no other solution.

---

For instance consider the site involved, `spiegel.de`, it renders fine with uBO without surrendering using the `genericblock` option.

**Reply by ameshkov** [reply]
I see your point, I just mean that you didn't actually remove that tool, they still have it and can use it, it's just not as convenient and not as clear as $genericblock:
```
@@$domain=spiegel.de
||thiscanbesafelyblocked.com^$domain=spiegel.de,important
```

Moreover, check the full set of rules for Spiegel, besides `$genericblock` there is a bunch of rules that unblocks stuff, probably these rules are added because of the missing `$genericblock` support? Makes it harder to $badfilter this in uBO filters where you keep the scriptlet/redirect solution to this.
```
@@||doubleclick.net/gampad/*adPositionName%3Dsuperbanner_$xmlhttprequest,domain=spiegel.de
@@||doubleclick.net/getconfig$xmlhttprequest,domain=spiegel.de
@@||g.doubleclick.net/gampad/ads?*superbanner_3$xmlhttprequest,domain=spiegel.de
@@||g.doubleclick.net/gampad/ads?*wallpaper_1$xmlhttprequest,domain=spiegel.de
@@||g.doubleclick.net/gpt/pubads_impl_$script,domain=spiegel.de
@@||g.doubleclick.net/tag/js/gpt.js$domain=spiegel.de
@@||pagead2.googlesyndication.com/getconfig/sodar?$xmlhttprequest,domain=spiegel.de
@@||spiegel.de^$genericblock,generichide
```

**Reply by gorhill** [reply]
Those exceptions are explicit, while it's not clear what `genericblock` will end up excepting -- a `genericblock` in EasyList may end up unduly affecting the effect of EasyPrivacy.

I just think it's a bad filter option and made a choice to not implement it.. Consider there is only one single instance in EasyList, and no instance in EasyPrivacy.

I just prefer to not burden uBO with throwing code into it to support a dubious (in my opinion) filter option with little value as per usage in the real world. My goal is not to mimic exactly what other blockers do to the point where I would feel I have to burden uBO with supporting `genericblock`. Other blockers are free to support it -- and I am not trying to convince you to not support it, I am just explaining why I chose to not support it.

**Comment by ameshkov** [comment]
@mjethani some of the differences are due to missing `ping` mapping in `tsurlfilter.js`: `  [ 'ping', RequestType.Ping ],`

**Reply by mjethani** [reply]
@ameshkov see https://github.com/cliqz-oss/adblocker/pull/2171
